ms2rescore/parse_spectra.py

```diff
@@ -2,48 +2,148 @@
-        for run, psms in runs.items():
+        for run_name, psms in runs.items():
             psm_list_run = PSMList(psm_list=list(chain.from_iterable(psms.values())))
-            spectrum_file = infer_spectrum_path(config["spectrum_path"], run)
-            logger.debug("Reading spectrum file: '%s'", spectrum_file)
+            spectrum_file = infer_spectrum_path(spectrum_path, run_name)
+
+            LOGGER.debug("Reading spectrum file: '%s'", spectrum_file)
             precursors = get_precursor_info(str(spectrum_file))
 
-            if config["spectrum_id_pattern"]:
-                spectrum_id_pattern = re.compile(config["spectrum_id_pattern"])
+            # Parse spectrum IDs with regex pattern if provided
+            if spectrum_id_pattern:
+                compiled_pattern = re.compile(spectrum_id_pattern)
                 precursors = {
-                    spectrum_id_pattern.search(spectrum_id).group(1): precursor
+                    compiled_pattern.search(spectrum_id).group(1): precursor
                     for spectrum_id, precursor in precursors.items()
                 }
 
+            # Ensure all PSMs have a precursor values
             for psm in psm_list_run:
-                try:
-                    if rt_required:
-                        psm.retention_time = precursors[psm.spectrum_id].rt
-                    if im_required:
-                        psm.ion_mobility = precursors[psm.spectrum_id].im
-                    if not psm.precursor_mz:
-                        psm.precursor_mz = precursors[psm.spectrum_id].mz
-                except KeyError as e:
+                if psm.spectrum_id not in precursors:
                     raise SpectrumParsingError(
-                        f"Could not extract missing RT/IM values from spectrum file for run {run}."
-                    ) from e
+                        "Mismatch between PSM and spectrum file IDs. Could find precursor values "
+                        f"for PSM with ID {psm.spectrum_id} in run {run_name}.\n"
+                        "Please check that the `spectrum_id_pattern` and `psm_id_pattern` options "
+                        "are configured correctly. See "
+                        "https://ms2rescore.readthedocs.io/en/stable/userguide/configuration/#mapping-psms-to-spectra"
+                        " for more information.\n"
+                        f"Example ID from PSM file: {psm.spectrum_id}\n"
+                        f"Example ID from spectrum file: {list(precursors.keys())[0]}"
+                    )
+
+            # Store precursor values in dictionary
+            precursor_dict[run_name] = precursors
+
+    # Reshape precursor values into arrays matching PSM list
+    mzs = np.array([precursor_dict[psm.run][psm.spectrum_id].mz for psm in psm_list])
+    rts = np.array([precursor_dict[psm.run][psm.spectrum_id].rt for psm in psm_list])
+    ims = np.array([precursor_dict[psm.run][psm.spectrum_id].im for psm in psm_list])
```

Comment:
As a note, I considered if it would be worthwhile to use generators for this, to reduce memory consumption, and possibly some time spent on garbage collection. As such, it is slower, which is to be expected given the increased number of function calls. The question is though if the memory consumption for these is big enough?
```
In [23]: test = list(range(100000))

In [24]: %timeit np.fromiter((i for i in test), int)
5.44 ms ± 8.02 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [25]: %timeit np.array([i for i in test], int)
4.51 ms ± 34.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [26]: %timeit np.array([i for i in test])
5.14 ms ± 9.18 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```